ckan/model/user.py

```diff
@@ -1,37 +1,53 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
 import datetime
 import re
-import os
 from hashlib import sha1, md5
 
 import passlib.utils
 from passlib.hash import pbkdf2_sha512
 from sqlalchemy.sql.expression import or_
 from sqlalchemy.orm import synonym
 from sqlalchemy import types, Column, Table
+from sqlalchemy.dialects.postgresql import JSON
+from sqlalchemy.ext.mutable import MutableDict
 import vdm.sqlalchemy
 
-import meta
-import core
-import types as _types
-import domain_object
-
-
-user_table = Table('user', meta.metadata,
-        Column('id', types.UnicodeText, primary_key=True,
-               default=_types.make_uuid),
-        Column('name', types.UnicodeText, nullable=False, unique=True),
-        Column('openid', types.UnicodeText),
-        Column('password', types.UnicodeText),
-        Column('fullname', types.UnicodeText),
-        Column('email', types.UnicodeText),
-        Column('apikey', types.UnicodeText, default=_types.make_uuid),
-        Column('created', types.DateTime, default=datetime.datetime.now),
-        Column('reset_key', types.UnicodeText),
-        Column('about', types.UnicodeText),
-        Column('activity_streams_email_notifications', types.Boolean,
-            default=False),
-        Column('sysadmin', types.Boolean, default=False),
-        )
+from ckan.model import types as _types
+from ckan.model import meta, core, domain_object
+
+
+user_table = Table(
+    'user',
+    meta.metadata,
+    Column(
+        'id',
+        types.UnicodeText,
+        primary_key=True,
+        default=_types.make_uuid
+    ),
+    Column('name', types.UnicodeText, nullable=False, unique=True),
+    Column('openid', types.UnicodeText),
+    Column('password', types.UnicodeText),
+    Column('fullname', types.UnicodeText),
+    Column('email', types.UnicodeText),
+    Column('apikey', types.UnicodeText, default=_types.make_uuid),
+    Column('created', types.DateTime, default=datetime.datetime.now),
+    Column('reset_key', types.UnicodeText),
+    Column('about', types.UnicodeText),
+    Column(
+        'activity_streams_email_notifications',
+        types.Boolean,
+        default=False
+    ),
+    Column('sysadmin', types.Boolean, default=False),
+    # Internal user extras. This column can store any type that
+    # can be serialized to JSON. This isn't intended to be used
+    # for end-user key/values, but rather for extensions to store
+    # internal data, like a twitter username which it is then
+    # in charge of displaying.
+    Column('extras', MutableDict.as_mutable(JSON))
```

Comment:
#2670 is the PR that needs to get finished. Looks like we included a note in the readme on 2.5 but never included instructions for users upgrading.